docs/advanced/optimizing-selectors.md

```diff
@@ -0,0 +1,111 @@
+
+For example, there exists this state model:
+
+```ts
+interface SomeStateModel {
+  data: Data[];
+  name: string;
+}
+```
+
+And in this example there is an input component where a user can type a name. On key down, an action is dispatched updating the `name` property of state. On the same page, another component renders `data`. In order to render state data we create a selector in our state class:
+
+```ts
+@Selector()
+static getViewData(state: SomeStateModel) {
+   return state.data.map(d => expensiveFunction(d));
+}
+```
+
+Selectors defined in state classes implicitly have `state` injected as their first argument. The above selector will be recalculated every time the user types into the input component. Since `state` could update rapidly when a user types, the expensive selector will needlessly recalculate even though it does not care about the `name` property of `state` changing. This selector does not take advantage of memoization.
+
+One way to solve this problem is to turn off the `injectContainerState` selector [option](options.md) at root, state, or selector level. By default (in NGXS v3), the state is implicitly injected as the first argument for composite selectors _defined within state classes_. Turning off this setting prevents the container state from being injected as the first argument. This requires you to explicitly specify all arguments when you use the `@Selector([...])` decorator. Any parameterless `@Selector()` decorators will still inject the state as an implicit argument. Note that this option does not apply to selectors declared _outside of state classes_ (because there is no container state to inject). For example, we create two selectors in our state class:
+
+```ts
+@Selector([SomeState])
+static getData(state: SomeStateModel) {
+   return state.data;
+}
+
+@Selector([SomeState.getData])
+static getViewData(data: Data[]) {
+  return data.map(d => expensiveFunction(d));
+}
+```
+
+This `getViewData` selector will not be recalculated when a user types into the input component. This selector targets the specific property of `state` it cares about as its argument by leveraging an additional selector. When the `name` property of state changes, the `getViewData` arguments _do not change_. Memoization is taken advantage of.
+
+An alternative solution to turning off the selector option is to create a [meta selector](../concepts/select#meta-selectors). For example, we declare one selector in our state class and declare another selector outside of our state class:
+
+```ts
+@State({...})
+@Injectable()
+export class SomeState {
+  @Selector()
+  static getData(state: SomeStateModel) {
+    return state.data;
+  }
+}
+
+export class SomeSelectors {
```

Comment:
I really like this convention to suffix with `Queries` for these meta selectors.
What do you think?
```suggestion
export class SomeStateQueries {
```

Reply:
@arturovt also requested to change this class name (to what it is now).  I'm not sure there is a convention on the ngxs website.

On my team we do:

```export class ProductsSelectors```

Reply:
Hmmm... it is really a matter of preference.
I feel like the word "selector" gets really overused all over my codebase so I opted for something similar that can carry the meaning of the Q from CQRS ;-)
I don't feel strongly either way! 🤷‍♂️

Reply:
That definitely makes sense 👍